src/Collection/CollectionTrait.php

```diff
@@ -658,4 +658,24 @@ public function _unwrap()
     {
         return $this->unwrap();
     }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return \Cake\Collection
+     */
+    public function transpose()
+    {
+        $length = $this->map(function ($row) {
+            return count($row);
+        });
```

Comment:
You could just read the count of the first row. That should be enough, if you get a row with more or less elements that expected, throw an exception